# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent

### download and save locally
    with urllib.request.urlopen("https://data.bionet.nsw.gov.au/biosvcapp/odata/SpeciesNames",context=ssl.create_default_context(cafile=certifi.where())) as url:
        data = json.loads(url.read().decode())
    speciesnames = pd.json_normalize(data, record_path =['value'])
    speciesnames.to_csv(projectDir + 'source-data/nsw/nsw-species-names.csv', index=None)
    speciesnames

### 1. iNaturalist statuses

In [232]:
import pandas as pd
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
# taxastatus.head(3)
taxastatus

id taxon_id  user_id place_id source_id  \
0     166449    38493  1138587     7830             
1     234788   918383   702203     9994             
2     234789   918383   702203     7308             
3     166416  1033183  3669610     6825             
4     180721  1247288   222137     6825             
...      ...      ...      ...      ...       ...   
3448  165697  1182117  3669610    73684             
3449  130966   508985                               
3450  161226    50744   516268                      
3451  162783   924263   764897                      
3452  152695   779693   708886              16653   

                                              authority  \
0                    Flora and Fauna Guarantee Act 1988   
1                             Atlas of Living Australia   
2                             Atlas of Living Australia   
3                  NSW Office of Environment & Heritage   
4           NSW Threatened Species Scientific Committee   
...                                                 ...   
3448                              Australian Government   
3449                                                      
3450          DAWE Species Profile and Threats Database   
3451  Department of Biodiversity, Conservation and A...   
3452                                           QLD DEHP   

                     status  \
0                        CR   
1                       NT    
2                        LC   
3                        EN   
4                        vu   
...                     ...   
3448                     CR   
3449  critically endangered   
3450                     CR   
3451                     EN   
3452             endangered   

                                                    url description  \
0                                                                     
1     https://bie.ala.org.au/species/https://id.biod...               
2     https://bie.ala.org.au/species/https://id.biod...               
3     https://www.environment.nsw.gov.au/threateneds...               
4     https://www.environment.nsw.gov.au/topics/anim...               
...                                                 ...         ...   
3448  http://www.environment.gov.au/biodiversity/thr...               
3449  http://environment.gov.au/cgi-bin/sprat/public...               
3450  https://www.environment.gov.au/cgi-bin/sprat/p...               
3451  https://www.dpaw.wa.gov.au/images/documents/pl...               
3452  https://data.qld.gov.au/dataset/conservation-s...               

     geoprivacy  ...          genus specificEpithet infraspecificEpithet  \
0      obscured  ...      Eulamprus      kosciuskoi                        
1                ...   Chiloschista     phyllorhiza                        
2                ...   Chiloschista     phyllorhiza                        
3      obscured  ...       Eidothea      hardeniana                        
4      obscured  ...     Pomaderris         bodalla                        
...         ...  ...            ...             ...                  ...   
3448             ...    Achyranthes    margaretarum                        
3449   obscured  ...  Lichenostomus        melanops             cassidix   
3450   obscured  ...                                                       
3451   obscured  ...         Reedia       spathacea                        
3452   obscured  ...   Selenocosmia       crassipes                        

                  modified                   scientificName   taxonRank  \
0     2021-03-01T10:35:01Z             Eulamprus kosciuskoi     species   
1     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
2     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
3     2021-02-22T07:21:17Z              Eidothea hardeniana     species   
4     2021-08-27T06:18:35Z               Pomaderris bodalla     species   
...                    ...                              ...         ...   


In [233]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("NSW|New South Wales", ".nsw.")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
3       166416  1033183  3669610     6825             
2656    165059  1054498    58320     6825             
477     264941  1061113  3669610     6825             
371     180988  1070573  3669610     6825             
2514    160581  1076814   990532     6825             
...        ...      ...      ...      ...       ...   
1111    152247   906670   708886     6825     16650   
2637    164342   913024    58320     6825             
7       167825   953250   702203     6825             
836     155144   966856   708886     6825     16650   
2651    164615   973457   222137     6825             

                                              authority      status  \
3                  NSW Office of Environment & Heritage          EN   
2656               NSW Office of Environment & Heritage  Vulnerable   
477   New South Wales Biodiversity Conservation Act ...          EX   
371                NSW Office of Environment & Heritage          EN   
2514               NSW Office of Environment & Heritage  Vulnerable   
...                                                 ...         ...   
1111               NSW Office of Environment & Heritage  Endangered   
2637               NSW Office of Environment & Heritage  Vulnerable   
7     New South Wales Office of Environment and Heri...          VU   
836                NSW Office of Environment & Heritage  endangered   
2651        NSW Threatened Species Scientific Committee  Endangered   

                                                    url       description  \
3     https://www.environment.nsw.gov.au/threateneds...                     
2656  https://www.environment.nsw.gov.au/threatenedS...                     
477   https://bie.ala.org.au/species/https://id.biod...  Presumed Extinct   
371   https://www.environment.nsw.gov.au/threateneds...                     
2514  https://www.environment.nsw.gov.au/threateneds...                     
...                                                 ...               ...   
1111  https://www.environment.nsw.gov.au/resources/a...                     
2637  https://www.environment.nsw.gov.au/threatenedS...                     
7     https://www.environment.nsw.gov.au/threatenedS...        Vulnerable   
836   https://www.environment.nsw.gov.au/resources/a...                     
2651  https://www.environment.nsw.gov.au/topics/anim...                     

     geoprivacy  ...         genus specificEpithet infraspecificEpithet  \
3      obscured  ...      Eidothea      hardeniana                        
2656   obscured  ...  Prostanthera  cryptandroides       cryptandroides   
477        open  ...        Leuzea       australis                        
371              ...        Rotala      tripartita                        
2514   obscured  ...   Pterostylis       nigricans                        
...         ...  ...           ...             ...                  ...   
1111   obscured  ...     Caladenia        concolor                        
2637             ...     Symplocos     baeuerlenii                        
7                ...     Pultenaea          glabra                        
836    obscured  ...        Pteris   platyzomopsis                        
2651   obscured  ...       Calotis       pubescens                        

                  modified                              scientificName  \
3     2021-02-22T07:21:17Z                         Eidothea hardeniana   
2656  2020-12-17T03:53:45Z  Prostanthera cryptandroides cryptandroides   
477   2022-06-11T11:38:00Z                            Leuzea australis   
371   2021-09-21T05:48:55Z                           Rotala tripartita   
2514  2022-07-12T14:19:22Z                       Pterostylis nigricans   
...                    ...                                         ...   
1111  2022-06-10T07:36:31Z                          Caladenia concolor   
2637  2022-03-14T15:27:03Z                       Symplocos baeuerlenii   

### 2. iNaturalist taxonomy

In [234]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

Sensitive list: `geoprivacy` = `obscured`
overrides


In [235]:
# %%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
# Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV
#
# sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr487?max=10000&includeKVP=true")
# sensitivelist = lf.kvp_to_columns(sensitivelist)
# sensitivelist.to_csv(sourcedir + "nsw-ala-sensitive.csv", index=False)
#
# conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr650?max=10000&includeKVP=true")
# conservationlist = lf.kvp_to_columns(conservationlist)
# conservationlist.to_csv(sourcedir + "nsw-ala-conservation.csv", index=False)

In [236]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "nsw-ala-sensitive.csv", dtype=str)
sensitivelist = sensitivelist.rename(columns={'TSProfileID':'nsw_taxonID'})
sensitivelist = sensitivelist.rename(columns={'status':'nsw_status'})
sensitivelist = sensitivelist.rename(columns={'scientificName':'nsw_scientificName'})
sensitivelist['bionet_geoprivacy'] = 'obscured'
sensitivelist

taxonRank   kingdom     class           order       family  \
0       Species  Animalia      Aves  Psittaciformes   Cacatuidae   
1       Species  Animalia      Aves    Strigiformes    Strigidae   
2       Species  Animalia      Aves    Strigiformes    Tytonidae   
3       Species  Animalia      Aves    Strigiformes    Strigidae   
4       Species  Animalia  Reptilia        Squamata     Elapidae   
..          ...       ...       ...             ...          ...   
194  Subspecies   Plantae     Flora           Flora      Poaceae   
195     Species   Plantae     Flora           Flora  Orchidaceae   
196     Species   Plantae     Flora           Flora  Orchidaceae   
197     Species   Plantae     Flora           Flora  Orchidaceae   
198     Species   Plantae     Flora           Flora    Myrtaceae   

             genus                    nsw_scientificName specificEpithet  \
0    Callocephalon              Callocephalon fimbriatum      fimbriatum   
1            Ninox                         Ninox strenua         strenua   
2             Tyto                      Tyto tenebricosa     tenebricosa   
3            Ninox                       Ninox connivens       connivens   
4    Hoplocephalus             Hoplocephalus bungaroides     bungaroides   
..             ...                                   ...             ...   
194    Anthosachne  Anthosachne kingiana subsp. kingiana        kingiana   
195      Caladenia                       Caladenia tensa           tensa   
196      Caladenia                  Caladenia atroclavia      atroclavia   
197   Prasophyllum                  Prasophyllum sandrae         sandrae   
198     Backhousia                Backhousia subargentea     subargentea   

                  vernacularName    establishmentMeans  ... protectedInNSW  \
0             Gang-gang Cockatoo  Alive in NSW, Native  ...           true   
1                   Powerful Owl  Alive in NSW, Native  ...           true   
2                      Sooty Owl  Alive in NSW, Native  ...           true   
3                    Barking Owl  Alive in NSW, Native  ...           true   
4             Broad-headed Snake  Alive in NSW, Native  ...           true   
..                           ...                   ...  ...            ...   
194    Philip Island Wheat Grass  Alive in NSW, Native  ...          false   
195          Rigid Spider-orchid  Alive in NSW, Native  ...           true   
196  Black-clubbed Spider-orchid  Alive in NSW, Native  ...           true   
197     Majors Creek Leek Orchid  Alive in NSW, Native  ...          false   
198               Giant Ironwood  Alive in NSW, Native  ...          false   

    sensitivityClass nsw_taxonID    countryConservation  \
0         Category 3     10975.0             Endangered   
1         Category 3     10562.0             Not Listed   
2         Category 3     10821.0             Not Listed   
3         Category 3     10561.0             Not Listed   
4         Category 2     10413.0             Vulnerable   
..               ...         ...                    ...   
194       Category 3     20148.0  Critically Endangered   
195       Category 2         NaN             Not Listed   
196       Category 2     20365.0             Endangered   
197       Category 2     10668.0             Not Listed   
198       Category 3     10166.0             Not Listed   

                  dcterms_modified speciesID taxonID generalisation  \
0     2022-05-12T15:50:01.01+10:00         6       6            1km   
1        2011-11-11T11:23:06+11:00       162     162            1km   
2    2012-07-23T18:04:52.683+10:00       333     333            1km   
3        2011-11-11T11:23:06+11:00       363     363            1km   
4        2011-11-11T11:23:04+11:00       390     390           10km   
..                             ...       ...     ...            ...   
194   2020-01-23T17:14:13.01+11:00     23014   23014            1km   
195   2020-06-23T19:56:43.85+10:00     23737   23737          

In [237]:
# Fix duplicate detected
sensitivelist.at[145, 'bionet_geoprivacy'] = 'private'
sensitivelist.iloc[145]

taxonRank                                   Species
kingdom                                     Plantae
class                                         Flora
order                                         Flora
family                                Araucariaceae
genus                                      Wollemia
nsw_scientificName                 Wollemia nobilis
specificEpithet                             nobilis
vernacularName                         Wollemi Pine
establishmentMeans             Alive in NSW, Native
sourceStatus                  Critically Endangered
protectedInNSW                                 true
sensitivityClass                         Category 1
nsw_taxonID                                 10840.0
countryConservation           Critically Endangered
dcterms_modified       2018-10-26T23:43:22.22+11:00
speciesID                                     13940
taxonID                                       13940
generalisation                             WITHHOLD
nsw_status  

In [238]:
# Read conservation list data
# bionet geoprivacy conditions to apply for conservation list
def gpcondition(x):
    if x=='Category 1':
        return "private"
    elif x== 'Category 2' or x=='Category 3':
        return "obscured"
    else:
        return 'open'
conservationlist = pd.read_csv(sourcedir + "nsw-ala-conservation.csv", dtype=str)
conservationlist = conservationlist.rename(columns={'tsprofileid':'nsw_taxonID'})
conservationlist = conservationlist.rename(columns={'status':'nsw_status'})
conservationlist = conservationlist.rename(columns={'scientificName':'nsw_scientificName'})
conservationlist['bionet_geoprivacy'] = conservationlist['sensitivityClass'].apply(gpcondition)
conservationlist

taxonRank   kingdom         class            order          family  \
0        Species  Animalia      Reptilia         Squamata     Pygopodidae   
1        Species  Animalia          Aves   Psittaciformes      Cacatuidae   
2        Species  Animalia      Reptilia         Squamata        Elapidae   
3        Species  Animalia      Amphibia            Anura         Hylidae   
4        Species  Animalia          Aves    Passeriformes    Meliphagidae   
...          ...       ...           ...              ...             ...   
1020     Species   Plantae         Flora            Flora  Pittosporaceae   
1021  Subspecies  Animalia          Aves    Passeriformes       Maluridae   
1022     Species  Animalia       Insecta       Coleoptera        Cleridae   
1023     Species  Animalia  Malacostraca        Amphipoda        Hyalidae   
1024     Species  Animalia      Mammalia  Peramelemorphia     Peramelidae   

              genus             nsw_scientificName    specificEpithet  \
0             Delma                    Delma impar              impar   
1     Callocephalon       Callocephalon fimbriatum         fimbriatum   
2          Cacophis            Cacophis harriettae         harriettae   
3           Litoria        Litoria booroolongensis    booroolongensis   
4       Anthochaera            Anthochaera phrygia            phrygia   
...             ...                            ...                ...   
1020    Pittosporum  Pittosporum sp. Coffs Harbour  sp. Coffs Harbour   
1021      Amytornis       Amytornis striatus howei           striatus   
1022       Cormodes               Cormodes darwini            darwini   
1023        Litoria                Litoria watsoni            watsoni   
1024      Perameles             Perameles fasciata           fasciata   

                         vernacularName    establishmentMeans  \
0                Striped Legless Lizard  Alive in NSW, Native   
1                    Gang-gang Cockatoo  Alive in NSW, Native   
2                   White-crowned Snake  Alive in NSW, Native   
3                       Booroolong Frog  Alive in NSW, Native   
4                     Regent Honeyeater  Alive in NSW, Native   
...                                 ...                   ...   
1020                                NaN  Alive in NSW, Native   
1021                                NaN  Alive in NSW, Native   
1022                                NaN  Alive in NSW, Native   
1023                                NaN  Alive in NSW, Native   
1024  Liverpool Plains Barred Bandicoot  Alive in NSW, Native   

               sourceStatus protectedInNSW sensitivityClass nsw_taxonID  \
0                Vulnerable           true    Not Sensitive     10211.0   
1                Vulnerable           true       Category 3     10975.0   
2                Vulnerable           true    Not Sensitive     10117.0   
3                Endangered           true    Not Sensitive     10484.0   
4     Critically Endangered           true    Not Sensitive     10841.0   
...                     ...            ...              ...         ...   
1020  Critically Endangered          false    Not Sensitive     20378.0   
1021             Endangered          false    Not Sensitive     20422.0   
1022  Critically Endangered          false    Not Sensitive     20424.0   
1023             Endangered          false    Not Sensitive     20425.0   
1024                Extinct          false    Not Sensitive     20188.0   

        countryConservation               dcterms_modified speciesID taxonID  \
0                Vulnerable   2013-12-20T15:06:07.13+11:00         1       1   
1                Endangered   2022-05-12T15:50:01.01+10:00         6       6   
2                Not Listed  2015-05-28T10:14:29.083+10:00        31      31   
3                Endangered      2011-11-11T11:23:05+11:00        33      33   
4     Critically Endangered      2011-11-11T11:23:09+11:00        53      53   
...                     ...                

In [239]:
conservationlist.columns


Index(['taxonRank', 'kingdom', 'class', 'order', 'family', 'genus',
       'nsw_scientificName', 'specificEpithet', 'vernacularName',
       'establishmentMeans', 'sourceStatus', 'protectedInNSW',
       'sensitivityClass', 'nsw_taxonID', 'countryConservation',
       'dcterms_modified', 'speciesID', 'taxonID', 'nsw_status',
       'bionet_geoprivacy'],
      dtype='object')

In [240]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['nsw_taxonID','nsw_scientificName','nsw_status','bionet_geoprivacy','kingdom','class','family']],
                    conservationlist[['nsw_taxonID','nsw_scientificName','nsw_status','bionet_geoprivacy','kingdom','class','family']]]).drop_duplicates()
statelist['nsw_taxonID'] = statelist['nsw_taxonID'].str[:-2]
# retrieve binomial and trinomial names from GBIF
statelist = statelist.rename(columns={'nsw_scientificName':'name'})
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "nsw-gbif.csv", index=False)
statelist = statelist.merge(parsednames[['scientificName','canonicalName','canonicalNameComplete','type','rankMarker']],how="left",left_on="name",right_on="scientificName")
numfullstatelist = len(statelist.index)
conservationlist = conservationlist.rename(columns={'name':'nsw_scientificName'})
statelist['nsw_scientificName'] = statelist['canonicalName']
statelist

nsw_taxonID                           name             nsw_status  \
0          10975       Callocephalon fimbriatum             Vulnerable   
1          10562                  Ninox strenua             Vulnerable   
2          10821               Tyto tenebricosa             Vulnerable   
3          10561                Ninox connivens             Vulnerable   
4          10413      Hoplocephalus bungaroides             Endangered   
...          ...                            ...                    ...   
1026       20378  Pittosporum sp. Coffs Harbour  Critically Endangered   
1027       20422       Amytornis striatus howei             Endangered   
1028       20424               Cormodes darwini  Critically Endangered   
1029       20425                Litoria watsoni             Endangered   
1030       20188             Perameles fasciata                Extinct   

     bionet_geoprivacy   kingdom         class          family  \
0             obscured  Animalia          Aves      Cacatuidae   
1             obscured  Animalia          Aves       Strigidae   
2             obscured  Animalia          Aves       Tytonidae   
3             obscured  Animalia          Aves       Strigidae   
4             obscured  Animalia      Reptilia        Elapidae   
...                ...       ...           ...             ...   
1026              open   Plantae         Flora  Pittosporaceae   
1027              open  Animalia          Aves       Maluridae   
1028              open  Animalia       Insecta        Cleridae   
1029              open  Animalia  Malacostraca        Hyalidae   
1030              open  Animalia      Mammalia     Peramelidae   

                     scientificName              canonicalName  \
0          Callocephalon fimbriatum   Callocephalon fimbriatum   
1                     Ninox strenua              Ninox strenua   
2                  Tyto tenebricosa           Tyto tenebricosa   
3                   Ninox connivens            Ninox connivens   
4         Hoplocephalus bungaroides  Hoplocephalus bungaroides   
...                             ...                        ...   
1026  Pittosporum sp. Coffs Harbour       Pittosporum sp.Coffs   
1027       Amytornis striatus howei   Amytornis striatus howei   
1028               Cormodes darwini           Cormodes darwini   
1029                Litoria watsoni            Litoria watsoni   
1030             Perameles fasciata         Perameles fasciata   

          canonicalNameComplete        type rankMarker  \
0      Callocephalon fimbriatum  SCIENTIFIC        sp.   
1                 Ninox strenua  SCIENTIFIC        sp.   
2              Tyto tenebricosa  SCIENTIFIC        sp.   
3               Ninox connivens  SCIENTIFIC        sp.   
4     Hoplocephalus bungaroides  SCIENTIFIC        sp.   
...                         ...         ...        ...   
1026       Pittosporum sp.Coffs    INFORMAL        sp.   
1027   Amytornis striatus howei  SCIENTIFIC   infrasp.   
1028           Cormodes darwini  SCIENTIFIC        sp.   
1029            Litoria watsoni  SCIENTIFIC        sp.   
1030         Perameles fasciata  SCIENTIFIC        sp.   

             nsw_scientificName  
0      Callocephalon fimbriatum  
1                 Ninox strenua  
2              Tyto tenebricosa  
3               Ninox connivens  
4     Hoplocephalus bungaroides  
...                         ...  
1026       Pittosporum sp.Coffs  
1027   Amytornis striatus howei  
1028           Cormodes darwini  
1029            Litoria watsoni  
1030         Perameles fasciata  

[1031 rows x 13 columns]

In [241]:
# Identify records that won't comply with iNaturalist species names
noncomply = statelist[statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID', 'BLACKLISTED']) ]

In [242]:
# remove records that do not comply
statelist = statelist[~statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID', 'BLACKLISTED']) ]
statelist = pd.DataFrame(statelist[['nsw_taxonID','nsw_scientificName','nsw_status','bionet_geoprivacy']]).drop_duplicates()
statelist

nsw_taxonID         nsw_scientificName             nsw_status  \
0          10975   Callocephalon fimbriatum             Vulnerable   
1          10562              Ninox strenua             Vulnerable   
2          10821           Tyto tenebricosa             Vulnerable   
3          10561            Ninox connivens             Vulnerable   
4          10413  Hoplocephalus bungaroides             Endangered   
...          ...                        ...                    ...   
1025       10521       Melichrus gibberagee             Endangered   
1027       20422   Amytornis striatus howei             Endangered   
1028       20424           Cormodes darwini  Critically Endangered   
1029       20425            Litoria watsoni             Endangered   
1030       20188         Perameles fasciata                Extinct   

     bionet_geoprivacy  
0             obscured  
1             obscured  
2             obscured  
3             obscured  
4             obscured  
...                ...  
1025              open  
1027              open  
1028              open  
1029              open  
1030              open  

[1010 rows x 4 columns]

In [243]:
parsednames['type'].unique()

#cols for debugging

array(['SCIENTIFIC', 'INFORMAL', 'BLACKLISTED'], dtype=object)

In [244]:
# check for duplicates with conflicting information
dupinformation = statelist.groupby('nsw_taxonID').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
dupinformation
#df.groupby('hash').filter(lambda group: len(group) > 1).sort('size', ascending=False)

Empty DataFrame
Columns: [nsw_taxonID, nsw_scientificName, nsw_status, bionet_geoprivacy]
Index: []

In [245]:
# fix duplicate

### 4. Equivalent IUCN statuses

In [246]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['nsw_status'])['nsw_status'].count()

nsw_status
Critically Endangered    111
Endangered               418
Extinct                   72
Not Listed                 6
Vulnerable               403
Name: nsw_status, dtype: int64

In [247]:
iucnStatusMappings = {
    'critically endangered':'Critically Endangered',
    'endangered':'Endangered',
    'extinct':'Extinct',
    'not listed': 'Vulnerable',
    'vulnerable': 'Vulnerable'
}


### 5. Determine best place ID to use

In [248]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6825

place_id  place_name       place_display_name 
6744      Australia        Australia                5
6825      New South Wales  New South Wales, AU    161
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [249]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['nsw_scientificName', 'nsw_taxonID','nsw_status','bionet_geoprivacy']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='nsw_scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses

nsw_scientificName nsw_taxonID             nsw_status  \
140               Acacia atrox       10003             Endangered   
621       Acacia baueri aspera       10005             Vulnerable   
97          Acacia dangarensis       20028  Critically Endangered   
606           Acacia pubifolia       10024             Endangered   
443              Acacia ruppii       10027             Endangered   
...                        ...         ...                    ...   
1008      Melichrus gibberagee       10521             Endangered   
1009  Amytornis striatus howei       20422             Endangered   
1010          Cormodes darwini       20424  Critically Endangered   
1011           Litoria watsoni       20425             Endangered   
1012        Perameles fasciata       20188                Extinct   

     bionet_geoprivacy status_id        scientificName taxon_id user_id  \
140           obscured    152295          Acacia atrox   898643  708886   
621               open    159920  Acacia baueri aspera  1111698  702203   
97            obscured    152286    Acacia dangarensis   775137  708886   
606               open    162893      Acacia pubifolia   897779  702203   
443               open    162887         Acacia ruppii   578518  702203   
...                ...       ...                   ...      ...     ...   
1008              open       NaN                   NaN      NaN     NaN   
1009              open       NaN                   NaN      NaN     NaN   
1010              open       NaN                   NaN      NaN     NaN   
1011              open       NaN                   NaN      NaN     NaN   
1012              open       NaN                   NaN      NaN     NaN   

     description iucn                             authority      status  \
140                40  NSW Office of Environment & Heritage  endangered   
621                30                       New South Wales          VU   
97                 40  NSW Office of Environment & Heritage  endangered   
606                40                       New South Wales          EN   
443                40                       New South Wales          EN   
...          ...  ...                                   ...         ...   
1008         NaN  NaN                                   NaN         NaN   
1009         NaN  NaN                                   NaN         NaN   
1010         NaN  NaN                                   NaN         NaN   
1011         NaN  NaN                                   NaN         NaN   
1012         NaN  NaN                                   NaN         NaN   

     geoprivacy place_id   place_display_name  
140    obscured     6825  New South Wales, AU  
621                 6825  New South Wales, AU  
97     obscured     6825  New South Wales, AU  
606                 6825  New South Wales, AU  
443                 6825  New South Wales, AU  
...         ...      ...                  ...  
1008        NaN      NaN                  NaN  
1009        NaN      NaN                  NaN  
1010        NaN      NaN                  NaN  
1011        NaN      NaN                  NaN  
1012        NaN      NaN                  NaN  

[1013 rows x 15 columns]

In [250]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "New South Wales Office of Environment and Heritage"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://www.environment.nsw.gov.au/threatenedspeciesapp"
baseurl = "https://www.environment.nsw.gov.au/threatenedspeciesapp/profile.aspx?id="
mergedstatuses['new_url'] = baseurl + mergedstatuses['nsw_taxonID']
mergedstatuses.rename(columns={'bionet_geoprivacy':'new_geoprivacy'},inplace=True)
mergedstatuses['new_place_id'] = '6825'  # NEW SOUTH WALES
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

nsw_scientificName nsw_taxonID             nsw_status  \
140               Acacia atrox       10003             Endangered   
621       Acacia baueri aspera       10005             Vulnerable   
97          Acacia dangarensis       20028  Critically Endangered   
606           Acacia pubifolia       10024             Endangered   
443              Acacia ruppii       10027             Endangered   
...                        ...         ...                    ...   
1008      Melichrus gibberagee       10521             Endangered   
1009  Amytornis striatus howei       20422             Endangered   
1010          Cormodes darwini       20424  Critically Endangered   
1011           Litoria watsoni       20425             Endangered   
1012        Perameles fasciata       20188                Extinct   

     new_geoprivacy status_id        scientificName taxon_id user_id  \
140        obscured    152295          Acacia atrox   898643  708886   
621            open    159920  Acacia baueri aspera  1111698  702203   
97         obscured    152286    Acacia dangarensis   775137  708886   
606            open    162893      Acacia pubifolia   897779  702203   
443            open    162887         Acacia ruppii   578518  702203   
...             ...       ...                   ...      ...     ...   
1008           open       NaN                   NaN      NaN     NaN   
1009           open       NaN                   NaN      NaN     NaN   
1010           open       NaN                   NaN      NaN     NaN   
1011           open       NaN                   NaN      NaN     NaN   
1012           open       NaN                   NaN      NaN     NaN   

     description iucn  ... geoprivacy place_id   place_display_name  \
140                40  ...   obscured     6825  New South Wales, AU   
621                30  ...                6825  New South Wales, AU   
97                 40  ...   obscured     6825  New South Wales, AU   
606                40  ...                6825  New South Wales, AU   
443                40  ...                6825  New South Wales, AU   
...          ...  ...  ...        ...      ...                  ...   
1008         NaN  NaN  ...        NaN      NaN                  NaN   
1009         NaN  NaN  ...        NaN      NaN                  NaN   
1010         NaN  NaN  ...        NaN      NaN                  NaN   
1011         NaN  NaN  ...        NaN      NaN                  NaN   
1012         NaN  NaN  ...        NaN      NaN                  NaN   

                                          new_authority  \
140   New South Wales Office of Environment and Heri...   
621   New South Wales Office of Environment and Heri...   
97    New South Wales Office of Environment and Heri...   
606   New South Wales Office of Environment and Heri...   
443   New South Wales Office of Environment and Heri...   
...                                                 ...   
1008  New South Wales Office of Environment and Heri...   
1009  New South Wales Office of Environment and Heri...   
1010  New South Wales Office of Environment and Heri...   
1011  New South Wales Office of Environment and Heri...   
1012  New South Wales Office of Environment and Heri...   

                                        new_description  \
140   Listed as Threatened - refer to https://www.en...   
621   Listed as Threatened - refer to https://www.en...   
97    Listed as Threatened - refer to https://www.en...   
606   Listed as Threatened - refer to https://www.en...   
443   Listed as Threatened - refer to https://www.en...   
...                                                 ...   
1008  Listed as Threatened - refer to https://www.en...   
1009  Listed as Threatened - refer to https://www.en...   
1010  Listed as Threatened - refer to https://www.en...   
1011  Listed as Threatened - refer to https://www.en...   
1012  Listed as Threatened - refer to https://www.en...   

                                               

## Updates

In [251]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action             taxon_name      id taxon_id      status  \
140  UPDATE           Acacia atrox  152295   898643  endangered   
621  UPDATE   Acacia baueri aspera  159920  1111698          VU   
97   UPDATE     Acacia dangarensis  152286   775137  endangered   
606  UPDATE       Acacia pubifolia  162893   897779          EN   
443  UPDATE          Acacia ruppii  162887   578518          EN   
..      ...                    ...     ...      ...         ...   
2    UPDATE       Tyto tenebricosa  152244    20422  vulnerable   
112  UPDATE  Viola cleistogamoides  152321   566603  endangered   
143  UPDATE       Wollemia nobilis  152312    49381  Endangered   
702  UPDATE     Zieria involucrata  266172   604220          EN   
793  UPDATE       Zieria prostrata  165911   897289          EN   

    iucn_equivalent                                          authority  \
140      Endangered  New South Wales Office of Environment and Heri...   
621      Vulnerable  New South Wales Office of Environment and Heri...   
97       Endangered  New South Wales Office of Environment and Heri...   
606      Vulnerable  New South Wales Office of Environment and Heri...   
443      Vulnerable  New South Wales Office of Environment and Heri...   
..              ...                                                ...   
2        Vulnerable  New South Wales Office of Environment and Heri...   
112      Endangered  New South Wales Office of Environment and Heri...   
143      Endangered  New South Wales Office of Environment and Heri...   
702      Vulnerable  New South Wales Office of Environment and Heri...   
793      Vulnerable  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
140  https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
621  https://www.environment.nsw.gov.au/threateneds...       open     6825   
97   https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
606  https://www.environment.nsw.gov.au/threateneds...       open     6825   
443  https://www.environment.nsw.gov.au/threateneds...       open     6825   
..                                                 ...        ...      ...   
2    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
112  https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
143  https://www.environment.nsw.gov.au/threateneds...    private     6825   
702  https://www.environment.nsw.gov.au/threateneds...       open     6825   
793  https://www.environment.nsw.gov.au/threateneds...       open     6825   

      username                                        description  
140  peggydnew  Listed as Threatened - refer to https://www.en...  
621  peggydnew  Listed as Threatened - refer to https://www.en...  
97   peggydnew  Listed as Threatened - refer to https://www.en...  
606  peggydnew  Listed as Threatened - refer to https://www.en...  
443  peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
2    peggydnew  Listed as Threatened - refer to https://www.en...  
112  peggydnew  Listed as Threatened - refer to https://www.en...  
143  peggydnew  Listed as Threatened - refer to https://www.en...  
702  peggydnew  Listed as Threatened - refer to https://www.en...  
793  peggydnew  Listed as Threatened - refer to https://www.en...  

[149 rows x 12 columns]

## No status in iNaturalist via straight scientificName match
The NSW records that didn't match up to a status in iNaturalist

In [252]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="nsw_scientificName",right_on="scientificName")
noinatstatus

nsw_scientificName nsw_taxonID             nsw_status  \
0        Cyclopsitta diophthalma coxeni       10195  Critically Endangered   
1    Polytelis anthopeplus monarchoides       10644             Endangered   
2                  Pedionomus torquatus       10588             Endangered   
3                    Genoplesium baueri       10875             Endangered   
4                      Lindsaea fraseri       10481             Endangered   
..                                  ...         ...                    ...   
859                Melichrus gibberagee       10521             Endangered   
860            Amytornis striatus howei       20422             Endangered   
861                    Cormodes darwini       20424  Critically Endangered   
862                     Litoria watsoni       20425             Endangered   
863                  Perameles fasciata       20188                Extinct   

    new_geoprivacy status_id scientificName_x taxon_id user_id description  \
0         obscured       NaN              NaN      NaN     NaN         NaN   
1         obscured       NaN              NaN      NaN     NaN         NaN   
2         obscured       NaN              NaN      NaN     NaN         NaN   
3         obscured       NaN              NaN      NaN     NaN         NaN   
4         obscured       NaN              NaN      NaN     NaN         NaN   
..             ...       ...              ...      ...     ...         ...   
859           open       NaN              NaN      NaN     NaN         NaN   
860           open       NaN              NaN      NaN     NaN         NaN   
861           open       NaN              NaN      NaN     NaN         NaN   
862           open       NaN              NaN      NaN     NaN         NaN   
863           open       NaN              NaN      NaN     NaN         NaN   

    iucn  ...           class            order         family        genus  \
0    NaN  ...            Aves   Psittaciformes  Psittaculidae  Cyclopsitta   
1    NaN  ...            Aves   Psittaciformes  Psittaculidae    Polytelis   
2    NaN  ...            Aves  Charadriiformes   Pedionomidae   Pedionomus   
3    NaN  ...      Liliopsida      Asparagales    Orchidaceae  Genoplesium   
4    NaN  ...  Polypodiopsida     Polypodiales   Lindsaeaceae     Lindsaea   
..   ...  ...             ...              ...            ...          ...   
859  NaN  ...             NaN              NaN            NaN          NaN   
860  NaN  ...             NaN              NaN            NaN          NaN   
861  NaN  ...             NaN              NaN            NaN          NaN   
862  NaN  ...        Amphibia            Anura  Pelodryadidae      Litoria   
863  NaN  ...             NaN              NaN            NaN          NaN   

    specificEpithet infraspecificEpithet              modified  \
0       diophthalma               coxeni  2021-11-24T00:57:49Z   
1       anthopeplus         monarchoides  2019-09-08T01:35:11Z   
2         torquatus                  NaN  2022-06-11T05:20:23Z   
3            baueri                  NaN  2020-02-18T23:46:33Z   
4           fraseri                  NaN  2020-11-16T20:28:38Z   
..              ...                  ...                   ...   
859             NaN                  NaN                   NaN   
860             NaN                  NaN                   NaN   
861             NaN                  NaN                   NaN   
862         watsoni                  NaN  2020-10-26T20:46:33Z   
863             NaN                  NaN                   NaN   

                       scientificName_y   taxonRank  \
0        Cyclopsitta diophthalma coxeni  subspecies   
1    Polytelis anthopeplus monarchoides  subspecies   
2                  Pedionomus torquatus     species   
3                    Genoplesium baueri     species   
4                      Lindsaea fraseri     species   
..                                  ...         ...   
859                             

In [253]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

nsw_scientificName nsw_taxonID             nsw_status  \
0        Cyclopsitta diophthalma coxeni       10195  Critically Endangered   
1    Polytelis anthopeplus monarchoides       10644             Endangered   
2                  Pedionomus torquatus       10588             Endangered   
3                    Genoplesium baueri       10875             Endangered   
4                      Lindsaea fraseri       10481             Endangered   
..                                  ...         ...                    ...   
852                       Egernia roomi       20370  Critically Endangered   
853             Tympanocryptis osbornei       20380             Endangered   
854           Tympanocryptis mccartneyi       20379  Critically Endangered   
857                  Epacris gnidioides       10111             Vulnerable   
862                     Litoria watsoni       20425             Endangered   

    new_geoprivacy status_id scientificName_x taxon_id user_id description  \
0         obscured       NaN              NaN      NaN     NaN         NaN   
1         obscured       NaN              NaN      NaN     NaN         NaN   
2         obscured       NaN              NaN      NaN     NaN         NaN   
3         obscured       NaN              NaN      NaN     NaN         NaN   
4         obscured       NaN              NaN      NaN     NaN         NaN   
..             ...       ...              ...      ...     ...         ...   
852           open       NaN              NaN      NaN     NaN         NaN   
853           open       NaN              NaN      NaN     NaN         NaN   
854           open       NaN              NaN      NaN     NaN         NaN   
857           open       NaN              NaN      NaN     NaN         NaN   
862           open       NaN              NaN      NaN     NaN         NaN   

    iucn  ...           class            order         family           genus  \
0    NaN  ...            Aves   Psittaciformes  Psittaculidae     Cyclopsitta   
1    NaN  ...            Aves   Psittaciformes  Psittaculidae       Polytelis   
2    NaN  ...            Aves  Charadriiformes   Pedionomidae      Pedionomus   
3    NaN  ...      Liliopsida      Asparagales    Orchidaceae     Genoplesium   
4    NaN  ...  Polypodiopsida     Polypodiales   Lindsaeaceae        Lindsaea   
..   ...  ...             ...              ...            ...             ...   
852  NaN  ...        Reptilia         Squamata      Scincidae         Egernia   
853  NaN  ...        Reptilia         Squamata       Agamidae  Tympanocryptis   
854  NaN  ...        Reptilia         Squamata       Agamidae  Tympanocryptis   
857  NaN  ...   Magnoliopsida         Ericales      Ericaceae         Epacris   
862  NaN  ...        Amphibia            Anura  Pelodryadidae         Litoria   

    specificEpithet infraspecificEpithet              modified  \
0       diophthalma               coxeni  2021-11-24T00:57:49Z   
1       anthopeplus         monarchoides  2019-09-08T01:35:11Z   
2         torquatus                  NaN  2022-06-11T05:20:23Z   
3            baueri                  NaN  2020-02-18T23:46:33Z   
4           fraseri                  NaN  2020-11-16T20:28:38Z   
..              ...                  ...                   ...   
852           roomi                  NaN  2020-02-11T02:17:51Z   
853        osbornei                  NaN  2019-11-17T06:04:05Z   
854      mccartneyi                  NaN  2019-11-17T06:04:04Z   
857      gnidioides                  NaN  2020-02-09T06:28:44Z   
862         watsoni                  NaN  2020-10-26T20:46:33Z   

                       scientificName_y   taxonRank  \
0        Cyclopsitta diophthalma coxeni  subspecies   
1    Polytelis anthopeplus monarchoides  subspecies   
2                  Pedionomus torquatus     species   
3                    Genoplesium baueri     species   
4                      Lindsaea fraseri     species   
..                                  ...         ...

In [254]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['nsw_scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','nsw_scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'nsw_scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action                          taxon_name   id taxon_id      status  \
0      ADD      Cyclopsitta diophthalma coxeni  NaN   495917  Threatened   
1      ADD  Polytelis anthopeplus monarchoides  NaN   728998  Threatened   
2      ADD                Pedionomus torquatus  NaN     4341  Threatened   
3      ADD                  Genoplesium baueri  NaN   921754  Threatened   
4      ADD                    Lindsaea fraseri  NaN  1081779  Threatened   
..     ...                                 ...  ...      ...         ...   
852    ADD                       Egernia roomi  NaN  1031614  Threatened   
853    ADD             Tympanocryptis osbornei  NaN   965994  Threatened   
854    ADD           Tympanocryptis mccartneyi  NaN   965993  Threatened   
857    ADD                  Epacris gnidioides  NaN   918333  Threatened   
862    ADD                     Litoria watsoni  NaN  1145885  Threatened   

    iucn_equivalent                                          authority  \
0        Vulnerable  New South Wales Office of Environment and Heri...   
1        Vulnerable  New South Wales Office of Environment and Heri...   
2        Vulnerable  New South Wales Office of Environment and Heri...   
3        Vulnerable  New South Wales Office of Environment and Heri...   
4        Vulnerable  New South Wales Office of Environment and Heri...   
..              ...                                                ...   
852      Vulnerable  New South Wales Office of Environment and Heri...   
853      Vulnerable  New South Wales Office of Environment and Heri...   
854      Vulnerable  New South Wales Office of Environment and Heri...   
857      Vulnerable  New South Wales Office of Environment and Heri...   
862      Vulnerable  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
0    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
1    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
2    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
3    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
4    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
..                                                 ...        ...      ...   
852  https://www.environment.nsw.gov.au/threateneds...       open     6825   
853  https://www.environment.nsw.gov.au/threateneds...       open     6825   
854  https://www.environment.nsw.gov.au/threateneds...       open     6825   
857  https://www.environment.nsw.gov.au/threateneds...       open     6825   
862  https://www.environment.nsw.gov.au/threateneds...       open     6825   

      username                                        description  
0    peggydnew  Listed as Threatened - refer to https://www.en...  
1    peggydnew  Listed as Threatened - refer to https://www.en...  
2    peggydnew  Listed as Threatened - refer to https://www.en...  
3    peggydnew  Listed as Threatened - refer to https://www.en...  
4    peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
852  peggydnew  Listed as Threatened - refer to https://www.en...  
853  peggydnew  Listed as Threatened - refer to https://www.en...  
854  peggydnew  Listed as Threatened - refer to https://www.en...  
857  peggydnew  Listed as Threatened - refer to https://www.en...  
862  peggydnew  Listed as Threatened - refer to https://www.en...  

[743 rows x 12 columns]

### Write updates and additions to file

In [255]:
pd.concat([updates,additions]).to_csv(sourcedir + "nsw.csv", index=False)

In [256]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

nsw_scientificName nsw_taxonID             nsw_status  \
5       Callistemon linearifolius       10129             Vulnerable   
14           Gentiana bredboensis       10346  Critically Endangered   
16     Gentiana wingecarribiensis       10347  Critically Endangered   
27        Myriophyllum implicatum       20146  Critically Endangered   
28   Lysimachia vulgaris davurica       10498             Endangered   
..                            ...         ...                    ...   
858            Lobelia claviflora       20375  Critically Endangered   
859          Melichrus gibberagee       10521             Endangered   
860      Amytornis striatus howei       20422             Endangered   
861              Cormodes darwini       20424  Critically Endangered   
863            Perameles fasciata       20188                Extinct   

    new_geoprivacy status_id scientificName_x taxon_id user_id description  \
5         obscured       NaN              NaN      NaN     NaN         NaN   
14        obscured       NaN              NaN      NaN     NaN         NaN   
16        obscured       NaN              NaN      NaN     NaN         NaN   
27        obscured       NaN              NaN      NaN     NaN         NaN   
28        obscured       NaN              NaN      NaN     NaN         NaN   
..             ...       ...              ...      ...     ...         ...   
858           open       NaN              NaN      NaN     NaN         NaN   
859           open       NaN              NaN      NaN     NaN         NaN   
860           open       NaN              NaN      NaN     NaN         NaN   
861           open       NaN              NaN      NaN     NaN         NaN   
863           open       NaN              NaN      NaN     NaN         NaN   

    iucn  ... class order family genus specificEpithet infraspecificEpithet  \
5    NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
14   NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
16   NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
27   NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
28   NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
..   ...  ...   ...   ...    ...   ...             ...                  ...   
858  NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
859  NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
860  NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
861  NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   
863  NaN  ...   NaN   NaN    NaN   NaN             NaN                  NaN   

    modified scientificName_y taxonRank references  
5        NaN              NaN       NaN        NaN  
14       NaN              NaN       NaN        NaN  
16       NaN              NaN       NaN        NaN  
27       NaN              NaN       NaN        NaN  
28       NaN              NaN       NaN        NaN  
..       ...              ...       ...        ...  
858      NaN              NaN       NaN        NaN  
859      NaN              NaN       NaN        NaN  
860      NaN              NaN       NaN        NaN  
861      NaN              NaN       NaN        NaN  
863      NaN              NaN       NaN        NaN  

[121 rows x 38 columns]

### are there any that need to be removed?
nsw sensitive list count: 197
nsw inat statuses count: xx

updates to inat status: 152
additional inat status: 738
nsw statuses we can't find a taxon match for in iNaturalist: 139
total: 1395 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: xx-152=16 that may need checking

In [257]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
numnoncomply = len(noncomply.index)
numcomply = len(statelist.index)
numdupinfo = len(dupinformation.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Statelist merge': [numfullstatelist],
    'Species iNat Comply' : [numcomply],
    'Species iNat non-Comply': [numnoncomply],
    'Duplicate Information': [numdupinfo],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}
statsdf = pd.DataFrame(data=d)
statsdf


Sensitive  Conservation  Statelist merge  Species iNat Comply  \
0        199          1025             1031                 1010   

   Species iNat non-Comply  Duplicate Information  Updates  Additions  \
0                       21                      0      149        743   

   No Inat Status  Unknown to Inat  
0             864              121

In [258]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]

status_id taxon_id  user_id place_id source_id  \
477     264941  1061113  3669610     6825             
681     164757  1098159    58320     6825             
1147    152307   116844   708886     6825     16650   
736     166035  1194478  3669610     6825             
737     166036  1194478  3669610     6744             
51      167249  1227491  3669610     6825             
397     264070  1255425  3669610     6825             
809     262163  1289564   708886     6825     16650   
3367    152252    19111   708886     6825     16650   
2653    164637    19275    84719     6825             
2608    164300   369148   516268     6825             
1130    152282   505120   708886     6825     16650   
65      168033   634763   702203     6825             
1137    152293   775274   708886     6825     16650   
3426    152239   792286   708886     6825     16650   
2437    165525   906591    58320     6825             
836     155144   966856   708886     6825     16650   

                                              authority  \
477   New South Wales Biodiversity Conservation Act ...   
681                NSW Office of Environment & Heritage   
1147               NSW Office of Environment & Heritage   
736                NSW Office of Environment & Heritage   
737                               Australian Government   
51                 NSW Office of Environment & Heritage   
397                NSW Office of Environment & Heritage   
809                NSW Office of Environment & Heritage   
3367               NSW Office of Environment & Heritage   
2653                                      IUCN Red List   
2608        NSW Threatened Species Scientific Committee   
1130               NSW Office of Environment & Heritage   
65               Threatened Species Protection Act 1995   
1137               NSW Office of Environment & Heritage   
3426               NSW Office of Environment & Heritage   
2437               NSW Office of Environment & Heritage   
836                NSW Office of Environment & Heritage   

                     status  \
477                      EX   
681              Vulnerable   
1147  critically endangered   
736                      EN   
737                      EN   
51                       EN   
397                      CR   
809              vulnerable   
3367             vulnerable   
2653                     EN   
2608                     EN   
1130             vulnerable   
65                       EN   
1137  endangered population   
3426             vulnerable   
2437             Endangered   
836              endangered   

                                                    url  \
477   https://bie.ala.org.au/species/https://id.biod...   
681       https://www.inaturalist.org/taxa/1098159/edit   
1147  https://www.environment.nsw.gov.au/resources/a...   
736   https://www.environment.nsw.gov.au/threateneds...   
737   https://www.environment.nsw.gov.au/threateneds...   
51    https://www.environment.nsw.gov.au/threateneds...   
397   https://www.environment.nsw.gov.au/threateneds...   
809   https://www.environment.nsw.gov.au/resources/a...   
3367  https://www.environment.nsw.gov.au/resources/a...   
2653  https://www.iucnredlist.org/species/22727593/1...   
2608  https://www.dpi.nsw.gov.au/__data/assets/pdf_f...   
1130  https://www.environment.nsw.gov.au/resources/a...   
65    http://www.environment.gov.au/resource/nationa...   
1137  https://www.environment.nsw.gov.au/resources/a...   
3426  https://www.environment.nsw.gov.au/resources/a...   
2437  https://www.environment.nsw.gov.au/threatenedS...   
836   https://www.environment.nsw.gov.au/resources/a...   

                                            description geoprivacy  ...  \
477                                    Presumed Extinct       open  ...   
681                                                                 ...   
1147                                                      obscured  ...   
736                          